letsencrypt/plugins/standalone.py

```diff
@@ -0,0 +1,262 @@
+"""Standalone Authenticator."""
+import argparse
+import collections
+import functools
+import logging
+import random
+import socket
+import sys
+import threading
+
+import OpenSSL
+import six
+import zope.interface
+
+from acme import challenges
+from acme import crypto_util as acme_crypto_util
+from acme import standalone as acme_standalone
+
+from letsencrypt import achallenges
+from letsencrypt import errors
+from letsencrypt import interfaces
+
+from letsencrypt.plugins import common
+from letsencrypt.plugins import util
+
+logger = logging.getLogger(__name__)
+
+
+class ServerManager(object):
+    """Standalone servers manager.
+
+    Manager for `ACMEServer` and `ACMETLSServer` instances.
+
+    `certs` and `simple_http_resources` correspond to
+    `acme.crypto_util.SSLSocket.certs` and
+    `acme.crypto_util.SSLSocket.simple_http_resources` respectively. All
+    created servers share the same certificates and resources, so if
+    you're running both TLS and non-TLS instances, SimpleHTTP handlers
+    will serve the same URLs!
+
+    """
+    def __init__(self, certs, simple_http_resources):
+        self._servers = {}
+        self.certs = certs
+        self.simple_http_resources = simple_http_resources
+
+    def run(self, port, tls):
+        """Run ACME server on specified ``port``.
+
+        This method is idempotent, i.e. all calls with the same pair of
+        ``(port, tls)`` will reuse the same server.
+
+        :param int port: Port to run the server on.
+        :param bool tls: TLS or non-TLS?
+
+        :returns: Server instance (`ACMEServerMixin`) and the
+            corresponding (already started) thread (`threading.Thread`).
+        :rtype: tuple
+
+        """
+        if port in self._servers:
+            return self._servers[port]
+
+        logger.debug("Starting new server at %s (tls=%s)", port, tls)
+        handler = acme_standalone.ACMERequestHandler.partial_init(
+            self.simple_http_resources)
+
+        if tls:
+            cls = functools.partial(
+                acme_standalone.ACMETLSServer, certs=self.certs)
+        else:
+            cls = acme_standalone.ACMEServer
+
+        try:
+            server = cls(("", port), handler)
+        except socket.error as error:
+            raise errors.StandaloneBindError(error, port)
+
+        # if port == 0, then random free port on OS is taken
+        # pylint: disable=no-member
+        host, real_port = server.socket.getsockname()
+
+        thread = threading.Thread(target=server.serve_forever2)
+        logger.debug("Starting server at %s:%d", host, real_port)
+        thread.start()
+
+        self._servers[real_port] = (server, thread)
+        return self._servers[real_port]
```

Comment:
Bit of a nit, but I wonder if the `ServerManager` should be returning the running thread here and in `running()`. The `Thread` was ignored every time either function was called, which makes code slightly harder to read. Additionally, I feel like `ServerManager` should manage the threads it creates and I can't think of a reason off the top of my head why the caller needs the `Thread` object.